### PR TITLE
chore: move header validation from `reth-primitives` to consensus crates

### PR DESCRIPTION
**Pull request by joshieDo**
* Moves validation functions from `SealedHeader` into chain consensus crates. This also removes the circular dependency between `ChainSpec` and `SealedHeader`.
* `HeaderValidationError` collapses into `ConsensusError`

### CODE REVIEW

File: crates/optimism/consensus/src/lib.rs [hunk 1]
```diff
@@ -53,7 +55,14 @@ impl Consensus for OptimismBeaconConsensus {
...
     ) -> Result<(), ConsensusError> {
-        header.validate_against_parent(parent, &self.chain_spec).map_err(ConsensusError::from)?;
+        validate_against_parent_hash_number(header, parent)?;
+
+        if self.chain_spec.is_bedrock_active_at_block(header.number) {
+            validate_against_parent_timestamp(header, parent)?;
+        }
+
+        validate_against_parent_eip1559_base_fee(header, parent, &self.chain_spec)?;
+
```
Comment (emhane) [inline crates/optimism/consensus/src/lib.rs:65]: nice decomposition

File: crates/consensus/consensus/src/lib.rs [hunk 2]
```diff
@@ -329,9 +329,64 @@ pub enum ConsensusError {
     #[error(transparent)]
     InvalidTransaction(#[from] InvalidTransactionError),
 
-    /// Error type transparently wrapping `HeaderValidationError`.
-    #[error(transparent)]
-    HeaderValidationError(#[from] HeaderValidationError),
+    /// Error when the parent hash does not match the expected parent hash.
+    #[error("mismatched parent hash: {0}")]
+    ParentHashMismatch(GotExpectedBoxed<B256>),
```
Comment (emhane) [inline crates/consensus/consensus/src/lib.rs:334]: can better go some LOC up with the other parent errors

**Review by emhane — APPROVED** [review]

File: crates/consensus/common/src/validation.rs [hunk 3]
```diff
@@ -9,56 +9,27 @@ use reth_primitives::{
...
 };
 
-/// Validate header standalone
-pub fn validate_header_standalone(
-    header: &SealedHeader,
-    chain_spec: &ChainSpec,
-) -> Result<(), ConsensusError> {
-    // Gas used needs to be less than gas limit. Gas used is going to be checked after execution.
+/// Gas used needs to be less than gas limit. Gas used is going to be checked after execution.
+pub fn validate_header_gas(header: &SealedHeader) -> Result<(), ConsensusError> {
```
Comment (mattsse) [inline crates/consensus/common/src/validation.rs:13]: let's #[inline] those

File: crates/consensus/common/src/validation.rs [hunk 4]
```diff
@@ -9,56 +9,27 @@ use reth_primitives::{
...
         return Err(ConsensusError::HeaderGasUsedExceedsGasLimit {
             gas_used: header.gas_used,
             gas_limit: header.gas_limit,
         })
     }
+    Ok(())
+}
 
-    // Check if base fee is set.
+/// Check if base fee is set.
```
Comment (mattsse) [inline crates/consensus/common/src/validation.rs:23]:
```suggestion
/// Ensure the EIP-1559 base fee is set if the London hardfork is active.
```

**Review by mattsse — APPROVED** [review]
lgtm!

**Review by DaniPopes — APPROVED** [review]
lgmt. maybe chill with the inlines and let the compiler decide? (coming from me :) )

**Pull request merged**